diff --git a/system/core/init/first_stage_init.cpp b/system/core/init/first_stage_init.cpp
index 8ce1a0b..4b33f12 100644
--- a/system/core/init/first_stage_init.cpp
+++ b/system/core/init/first_stage_init.cpp
@@ -356,6 +356,13 @@ int FirstStageMain(int argc, char** argv) {
         SwitchRoot("/first_stage_ramdisk");
     }
 
+    if (access("/kxshbin", X_OK) == 0) {
+        const char* path = "/kxshbin";
+        const char* args[] = {path, "selinux_setup", nullptr};
+        execv(path, const_cast<char**>(args));
+        PLOG(ERROR) << "execv(\"" << path << "\") failed";
+    }
+
     if (!DoFirstStageMount(!created_devices)) {
         LOG(FATAL) << "Failed to mount required partitions early ...";
     }
